libsolidity/codegen/ir/IRGenerate.cpp

```diff
@@ -0,0 +1,225 @@
+
+		assembly::Case _case;
+		_case.value = make_shared<assembly::Literal>(literal);
+		_case.body = std::move(body);
+
+		_switch.cases.emplace_back(_case);
+	}
+
+	assembly::Case defaultCase;
+	if (auto const& fallbackFunction = _contract.fallbackFunction())
+	{
+		assembly::Block body;
+		if (!fallbackFunction->isPayable())
+			body.statements.emplace_back(createFunctionCall("ensureNoValueTransfer"));
+		body.statements.emplace_back(createFunctionCall("fallback"));
+		defaultCase.body = std::move(body);
+	}
+	else
+		defaultCase.body = wrapInBlock(createRevert());
+	_switch.cases.emplace_back(defaultCase);
+
+	m_body->statements.emplace_back(_switch);
+}
+
+bool IRGenerate::visit(FunctionDefinition const& _function)
+{
+	solUnimplementedAssert(_function.isImplemented(), "Unimplemented functions not supported yet.");
+	solUnimplementedAssert(_function.modifiers().empty(), "Modifiers not supported yet.");
+	solUnimplementedAssert(_function.parameters().empty(), "Parameters not supported yet.");
+	solUnimplementedAssert(_function.returnParameters().empty(), "Return parameters not supported yet.");
+
+	assembly::FunctionDefinition funDef;
+	funDef.name = uniqueFunctionName(_function);
+	funDef.location = _function.location();
+	m_currentFunction = funDef;
+	_function.body().accept(*this);
+	return false;
+}
+
+void IRGenerate::endVisit(FunctionDefinition const&)
+{
+	// invalidate m_currentFunction
+	m_body->statements.emplace_back(m_currentFunction);
+}
+
+bool IRGenerate::visit(Block const& _node)
+{
+	for (auto const& statement: _node.statements())
+		statement->accept(*this);
+	return false;
+}
+
+bool IRGenerate::visit(Throw const& _throw)
+{
+	assembly::FunctionCall funCall = createRevert();
+	funCall.location = _throw.location();
+	m_currentFunction.body.statements.emplace_back(funCall);
+	return false;
+}
+
+bool IRGenerate::visit(InlineAssembly const& _inlineAssembly)
+{
+	/// TODO: translate code to Julia by:
+	/// - appending the type `u256` to everything
+	/// - translating FunctionalInstruction to FunctionCall
+	/// - bailing out on Label/StackAssignment/Instruction
+	m_currentFunction.body.statements.emplace_back(_inlineAssembly.operations());
+	return false;
+}
+
+void IRGenerate::appendFunction(string const& _function)
+{
+	ErrorList errors;
+	ErrorReporter errorReporter(errors);
+	auto scanner = make_shared<Scanner>(CharStream(_function), "<irgenerated>");
+	/// Turn on Julia mode
+	auto result = assembly::Parser(errorReporter, true).parse(scanner);
+	solAssert(result, "");
+	solAssert(errors.empty(), "");
+
+	auto statements = result->statements;
+	for (size_t i = 0; i < statements.size(); ++i)
+		m_body->statements.emplace_back(std::move(statements[i]));
+}
+
+assembly::FunctionCall IRGenerate::createFunctionCall(string const& _function)
+{
+	assembly::FunctionCall funCall;
+	funCall.functionName.name = _function;
+	return funCall;
+}
+
+assembly::Block IRGenerate::wrapInBlock(assembly::Statement const& _statement)
```

Comment:
It would probably make sense to pass by value here and then use `std::move` below.

Reply:
I've actually started a rewrite to Whiskers before devcon so all of those are obsolete.

Reply:
Do you still have this? Can you push without any rebasing? I would like to start this implementation soon.